crates/macro-support/src/d.rs

```diff
@@ -0,0 +1,383 @@
+//! Structs to receive attribute options for each supported syntax type.
+//!
+//! Dedicated structs are used per-syntax-type to make sure unexpected meta properties produce errors
+//! at compile-time.
+
+use std::ops::Deref;
+
+use darling::{
+    util::{Flag, Override, SpannedValue},
+    Error, FromMeta,
+};
+use syn::{Ident, LitStr};
+use wasm_bindgen_backend::ast;
+
+use self::fields::SkipTypescript;
+
+/// A JS namespace, which can be represented as `"value"` or `r#"["first", "second"]#"`.
+pub struct JsNamespace(Vec<syn::LitStr>);
+
+impl JsNamespace {
+    pub fn to_strings(&self) -> Vec<String> {
+        self.0.iter().map(|s| s.value()).collect()
+    }
+}
+
+impl FromMeta for JsNamespace {
+    fn from_value(value: &syn::Lit) -> darling::Result<Self> {
+        if let syn::Lit::Str(s) = value {
+            if let Ok(array) = s.parse::<syn::ExprArray>() {
+                let mut errors = Error::accumulator();
+                let terms = array
+                    .elems
+                    .into_iter()
+                    .filter_map(|elem| errors.handle(get_string_literal(elem)))
+                    .collect();
+                errors.finish_with(Self(terms))
+            } else if let Ok(_ident) = s.parse::<syn::Ident>() {
+                Ok(Self(vec![s.clone()]))
+            } else {
+                Err(Error::custom(format!(
+                    "Value is not an ident or a list of idents"
+                )))
+            }
+        } else {
+            Err(Error::unexpected_lit_type(value))
+        }
+    }
+}
+
+fn get_string_literal(expr: syn::Expr) -> darling::Result<LitStr> {
+    if let syn::Expr::Lit(syn::ExprLit {
+        attrs: _,
+        lit: syn::Lit::Str(s),
+    }) = expr
+    {
+        Ok(s)
+    } else {
+        Err(darling::Error::custom("Expected string literal").with_span(&expr))
+    }
+}
+
+pub mod fields {
+    use darling::util::Flag;
+
+    macro_rules! field_trait {
+        ($name:ident, $return:ty, $field:ident) => {
+            pub trait $name {
+                fn $field(&self) -> $return;
+            }
+        };
+    }
+
+    field_trait!(JsName, Option<&syn::LitStr>, js_name);
+    field_trait!(SkipTypescript, Flag, skip_typescript);
+}
```

Comment:
I tried - and failed - to find a good way to make these driven by a single macro. It'd be very doable with a proc-macro, but having a `proc-macro` crate that supports `macro-support` ...which is a proc-macro... felt too recursive.